### ISSUE TITLE
`no-unnecessary-type-assertion` complains over valid code

### ISSUE THREAD

**Issue by jeffijoe** [palantir/tslint#3505]
### Bug Report

- __TSLint version__: 5.8.0
- __TypeScript version__: 2.6.1
- __Running TSLint via__: CLI

#### TypeScript code being linted

![image](https://user-images.githubusercontent.com/2505474/33073510-67328bfa-cec3-11e7-9d25-e21bcd19fc0d.png)

```ts
const { response } = await throws<AxiosError>(
  api.user1.createRole({ name: 'Nope' })
)
expect(response!.status).toBe(400)
expect(response!.data.message).toMatch(/admin/)
```

with `tslint.json` configuration:

```json
{
  "extends": ["tslint-config-standard", "tslint-config-prettier"],
  "rules": {
    "no-var-requires": false,
    "ordered-imports": [false],
    "no-console": [false],
    "member-access": [false],
    "object-literal-sort-keys": false,
    "await-promise": false,
    "no-use-before-declare": false
  }
}
```

#### Actual behavior

`no-unnecessary-type-assertion` complains over `response`, even though `response` is `AxiosResponse<any> | undefined`

```
ERROR: myfile.ts[13, 14]: This assertion is unnecessary since it does not change the type of the expression.
ERROR: myfile.ts[14, 14]: This assertion is unnecessary since it does not change the type of the expression.
```

#### Expected behavior

It should not error.

**Comment by ajafff** [comment]
@jeffijoe can you please try to make a create a minimal repro so I can debug this locally?

**Comment by jeffijoe** [comment]
@ajafff of course! I've attached a ZIP.

[tslint-repro.zip](https://github.com/palantir/tslint/files/1498875/tslint-repro.zip)

`npm install`, then open `src/test.ts` to see the correct code.

Now run `npm run lint` and see the correct code being broken by tslint.

**Comment by aluanhaddad** [comment]
I took a look. This is a downstream issue that is caused by the package `tslint-config-standard`. If I place the `no-unnecessary-type-assertion` rule directly in your tslint and remove the aforementioned package from `tslint.json` everything works as expected.

Moral of the story: Don't trust linter configs written by people who rely on ASI 😛.

**Comment by jeffijoe** [comment]
@aluanhaddad what is the issue with the `tslint-config-standard` package? What's making it "not work"? And what does ASI have to do with anything? 😛

**Comment by ajafff** [comment]
@aluanhaddad thanks for digging into this.

@jeffijoe This is actually caused by `no-unused-variable`. See #3455 for further details.

**Comment by aluanhaddad** [comment]
@jeffijoe I was just trying to get your goat :smiley:

**Comment by jeffijoe** [comment]
@ajafff I just read through that issue, not sure what the solution is? 😄

**Comment by ajafff** [comment]
@jeffijoe sorry, I picked the wrong of the dozens of similar issues. https://github.com/palantir/tslint/issues/3478#issuecomment-343699851 provides more information

**Comment by jeffijoe** [comment]
@ajafff thanks! :+1: Too bad workarounds are needed though. 😞

**Comment by buu700** [comment]
I'm running into a similar issue with this code: `const elements = <HTMLElement[]> Array.from(document.querySelectorAll('balls'));`, where the type of `elements` without the cast would have actually been `Element[]`, breaking code further down that relies on properties of `HTMLElement`.

Does this sound like the same root cause, @ajafff, or should I submit a new issue?

**Comment by ajafff** [comment]
@buu700 if you disable `no-unused-variable` and it works as expected, it's the same issue. Otherwise you should submit a new issue.

**Comment by buu700** [comment]
Got it, thanks! Just tested with `no-unnecessary-type-assertion` as the only rule enabled and reproduced the problem, so I'll submit a new issue.
